### PR TITLE
Do not use deprecated scala.deriving.ArrayProduct anymore.

### PR DESCRIPTION
**Pull request by sjrd**
Instead, wrap the array in a `Product` explicitly.

### CODE REVIEW

**Comment by lihaoyi** [comment]
I don't know much about the new Scala 3 APIs, @anatoliykmetyuk @jodersky could you review this change?

File: implicits/src-3/upickle/implicits/CaseClassReader.scala [hunk 1]
```diff
@@ -64,7 +64,13 @@ trait CaseClassReaderPiece extends MacrosCommon:
...
             throw new upickle.core.Abort("missing keys in dictionary: " + missingKeys.mkString(", "))
           }
-          m.fromProduct(ArrayProduct(values.toArray))
+
+          val valuesArray = values.toArray
+          m.fromProduct(new Product {
+            def canEqual(that: Any): Boolean = true
+            def productArity: Int = valuesArray.length
+            def productElement(i: Int): Any = valuesArray(i)
+          })
```
Comment (anatoliykmetyuk) [inline implicits/src-3/upickle/implicits/CaseClassReader.scala:73]: It does look like ArrayProduct was useful after all.

**Review by anatoliykmetyuk — APPROVED** [review]
LGTM provided that the removal of ArrayProduct from the API has strong arguments in favour.

File: implicits/src-3/upickle/implicits/CaseClassReader.scala (hunk 1 above)
Comment (sjrd) [reply implicits/src-3/upickle/implicits/CaseClassReader.scala:73]: I or we never said it wasn't useful at all. Just that its usefulness was too limited to deserve a place in the standard library, compared to all the other things that are not there and that could be useful too (even just for deriving).

**Comment by lihaoyi** [comment]
@anatoliykmetyuk I don't remember if I gave you publishing credentials, but feel free to tag/publish a new version if I did, and if I didn't give you credentials let me know if you'd like me to tag/publish a new version and I will do so

**Comment by anatoliykmetyuk** [comment]
I already published all the ported projects for M2. Getting this change in will have to wait until M3.

**Pull request merged**